### PR TITLE
Introduce soft and hard limits for memory limiter

### PR DESCRIPTION
**Pull request by tigrannajaryan**
Contributes to https://github.com/open-telemetry/opentelemetry-collector/issues/1121

**Description**

Memory limiter processor previously had only one limit. When exceeding
this limit it was previously continuously calling GC. This resulted in
huge CPU consumption if the check interval was small and this was
forcing to use large check intervals. This in turn was resulting in
lethargic response to growing memory usage and the memory limiter was
not very effective in situations when memory usage were growing rapidly
(e.g. when there was a big spike or when the backend was down).

I changed the logic of memory limiter to be based on 2 thresholds:
soft and hard. While below soft threshold the memory limiter is fully disabled.
Between soft and hard limiter the limiter begins dropping incoming data
but does not perform GC. Only when exceed the hard limit we perform GC.
The net result is that the actually used memory is limited at the level
set by soft limit and fluctuates between soft and hard limit as the garbage
is created and collected. Correspondingly GC runs much more infrequently, only
when the hard limit is reached and such GC immediately collects significant
amount of garbage (reduces memory usage close to soft limit) and thus does not
require subsequent GC calls for quite some time.

I did some performance tests with the old and new approaches with 4000 Mib limit,
100,000 spans per second and with exporter completely blocked (no backend).

With the old approach an interval of 100 ms causes about 450% of CPU usage
once the memory limit is hit (while below limit the CPU usage is around 50%).
Here is an extract of performance test output showing the moment when the limiter
is hit:
```
2020/12/03 20:20:47 Agent RAM (RES):3296 MiB, CPU:44.4% | Sent:   7022700 items | Received:         0 items (0/sec)
2020/12/03 20:20:50 Agent RAM (RES):3481 MiB, CPU:43.0% | Sent:   7322500 items | Received:         0 items (0/sec)
2020/12/03 20:20:53 Agent RAM (RES):3681 MiB, CPU:41.6% | Sent:   7614100 items | Received:         0 items (0/sec)
2020/12/03 20:20:56 Agent RAM (RES):3703 MiB, CPU:47.7% | Sent:   7863600 items | Received:         0 items (0/sec)
2020/12/03 20:20:59 Agent RAM (RES):3028 MiB, CPU:47.0% | Sent:   8062700 items | Received:         0 items (0/sec)
2020/12/03 20:21:02 Agent RAM (RES):3644 MiB, CPU:246.9% | Sent:   8331600 items | Received:         0 items (0/sec) <-- likely a regular GC, not at limit yet
2020/12/03 20:21:05 Agent RAM (RES):3555 MiB, CPU:72.8% | Sent:   8620500 items | Received:         0 items (0/sec)
2020/12/03 20:21:08 Agent RAM (RES):3717 MiB, CPU:57.5% | Sent:   8895500 items | Received:         0 items (0/sec)
2020/12/03 20:21:11 Agent RAM (RES):3877 MiB, CPU:126.9% | Sent:   9172900 items | Received:         0 items (0/sec) <-- hit limit
2020/12/03 20:21:14 Agent RAM (RES):3900 MiB, CPU:127.6% | Sent:   9461100 items | Received:         0 items (0/sec)
2020/12/03 20:21:17 Agent RAM (RES):3918 MiB, CPU:201.7% | Sent:   9728900 items | Received:         0 items (0/sec)
2020/12/03 20:21:20 Agent RAM (RES):3938 MiB, CPU:326.0% | Sent:   9994700 items | Received:         0 items (0/sec)
2020/12/03 20:21:23 Agent RAM (RES):3951 MiB, CPU:470.8% | Sent:  10253200 items | Received:         0 items (0/sec)
2020/12/03 20:21:26 Agent RAM (RES):3955 MiB, CPU:440.0% | Sent:  10504400 items | Received:         0 items (0/sec)
2020/12/03 20:21:29 Agent RAM (RES):3961 MiB, CPU:451.0% | Sent:  10766200 items | Received:         0 items (0/sec)
2020/12/03 20:21:32 Agent RAM (RES):3965 MiB, CPU:465.8% | Sent:  11008400 items | Received:         0 items (0/sec)
2020/12/03 20:21:35 Agent RAM (RES):3974 MiB, CPU:423.6% | Sent:  11272700 items | Received:         0 items (0/sec)
```
Even the interval of 1 second was unusable with the old approach and we had to
choose a longer interval to avoid performance degradation.

With the new approach under the exact same conditions when using 100ms check interval
the CPU usage is 50% when below memory limits and when the hard memory limits are
hit the CPU usage increases to 68%. With 1 second check interval there is no measurable
increase in CPU usage when memory limiter is hit (unlike 9x CPU increase with the old
approach).

Here is an extract of performance test output showing the moment when the limiter
is hit:
```
2020/12/03 20:28:35 Agent RAM (RES):1888 MiB, CPU:48.2% | Sent:   3796400 items | Received:         0 items (0/sec)
2020/12/03 20:28:38 Agent RAM (RES):2029 MiB, CPU:47.1% | Sent:   4088600 items | Received:         0 items (0/sec)
2020/12/03 20:28:41 Agent RAM (RES):2197 MiB, CPU:48.3% | Sent:   4388200 items | Received:         0 items (0/sec)
2020/12/03 20:28:44 Agent RAM (RES):2370 MiB, CPU:45.7% | Sent:   4679900 items | Received:         0 items (0/sec)
2020/12/03 20:28:47 Agent RAM (RES):2558 MiB, CPU:49.0% | Sent:   4972200 items | Received:         0 items (0/sec)
2020/12/03 20:28:50 Agent RAM (RES):2771 MiB, CPU:47.4% | Sent:   5260700 items | Received:         0 items (0/sec)
2020/12/03 20:28:53 Agent RAM (RES):2921 MiB, CPU:133.3% | Sent:   5547500 items | Received:         0 items (0/sec)
2020/12/03 20:28:56 Agent RAM (RES):2922 MiB, CPU:50.1% | Sent:   5846700 items | Received:         0 items (0/sec)
2020/12/03 20:28:59 Agent RAM (RES):2957 MiB, CPU:43.6% | Sent:   6131700 items | Received:         0 items (0/sec)
2020/12/03 20:29:02 Agent RAM (RES):3144 MiB, CPU:50.0% | Sent:   6419400 items | Received:         0 items (0/sec)
2020/12/03 20:29:05 Agent RAM (RES):3328 MiB, CPU:49.0% | Sent:   6719100 items | Received:         0 items (0/sec)
2020/12/03 20:29:08 Agent RAM (RES):3488 MiB, CPU:38.6% | Sent:   7007300 items | Received:         0 items (0/sec)
2020/12/03 20:29:11 Agent RAM (RES):3667 MiB, CPU:42.0% | Sent:   7306700 items | Received:         0 items (0/sec)
2020/12/03 20:29:14 Agent RAM (RES):3813 MiB, CPU:37.4% | Sent:   7577700 items | Received:         0 items (0/sec)
2020/12/03 20:29:17 Agent RAM (RES):3802 MiB, CPU:170.9% | Sent:   7860100 items | Received:         0 items (0/sec) <-- hit hard limit
2020/12/03 20:29:20 Agent RAM (RES):3882 MiB, CPU:68.1% | Sent:   8160000 items | Received:         0 items (0/sec)
2020/12/03 20:29:23 Agent RAM (RES):4007 MiB, CPU:42.3% | Sent:   8447900 items | Received:         0 items (0/sec)
2020/12/03 20:29:26 Agent RAM (RES):4007 MiB, CPU:39.3% | Sent:   8747800 items | Received:         0 items (0/sec)
2020/12/03 20:29:29 Agent RAM (RES):4008 MiB, CPU:34.3% | Sent:   9038400 items | Received:         0 items (0/sec)
2020/12/03 20:29:32 Agent RAM (RES):4009 MiB, CPU:39.9% | Sent:   9317200 items | Received:         0 items (0/sec)
```
This is a dramatically better picture compared to the old approach.

With 1 second interval memory limiter's impact on CPU is not measurable with the
new approach, whereas with the old approach it was still showing several times higher
CPU when limit was hit.

This makes small check intervals practically useful and allows to effectively
suppress incoming surges of data.

**Testing:** 
Tested manually, see above.

**Documentation:**
Changed README.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@kbrockhoff @jpkrohling it would be great to know your opinion on this.

File: processor/memorylimiter/memorylimiter.go
Comment (jpkrohling) [inline processor/memorylimiter/memorylimiter.go:243]: Just as an idea: if you use CompareAndSwap, you can simplify your `checkLimits` func (see comment there). Also, why `int64` and not perhaps `int32`?

File: processor/memorylimiter/memorylimiter.go
Comment (jpkrohling) [inline processor/memorylimiter/memorylimiter.go:269]: You can simplify this if your `setForcingDrop` compares the value before setting it, like this:

```go
forceDrop := ml.usageChecker.aboveSoftLimit(ms)
if ml.forcingDrop() && !forceDrop {
  // it would be a great to have this as a gauge!
  ml.logger.Info("Memory usage back within limits, resuming normal operation.", bytesToMib(ms))
}
if !ml.forcingDrop() && forceDrop {
  ml.logger.Warn("Memory usage is above soft limit, begin dropping data.", bytesToMib(ms))
}
ml.setForcingDrop(forceDrop)
```

File: processor/memorylimiter/memorylimiter.go
Comment (jpkrohling) [inline processor/memorylimiter/memorylimiter.go:312]: People looking at the metrics would thank you if you store this value (and the max) as a gauge (or expvar). This way, they wouldn't need to figure out where the config for this is located in their internal source repositories and which value should currently be used. This is specially true for values that aren't explicit, like this one here.

**Review by jpkrohling — COMMENTED** [review]
This looks great! I'll take some time later today, or early next week, to try this out under the same conditions as #1121.

File: processor/memorylimiter/memorylimiter.go
Comment (bogdandrutu) [inline processor/memorylimiter/memorylimiter.go:248]: The else branch can be removed from here.

File: processor/memorylimiter/memorylimiter.go
Comment (bogdandrutu) [reply processor/memorylimiter/memorylimiter.go:243]: No need for CAS, just a `SwapInt64` is enough for your idea.

File: processor/memorylimiter/memorylimiter.go
Comment (bogdandrutu) [reply processor/memorylimiter/memorylimiter.go:269]: I would do the set of the force dropping before logging, in case we are under pressure we want to start dropping before allocating more memory for logging.

File: processor/memorylimiter/memorylimiter.go
Comment (bogdandrutu) [inline processor/memorylimiter/memorylimiter.go:267]: before GC-ing I would set the dropping bit in case it was a jump from the non-dropping (less than soft) to greater than hard.

File: processor/memorylimiter/memorylimiter.go
Comment (bogdandrutu) [inline processor/memorylimiter/memorylimiter.go:252]: s/bytesToMib/zapCurrentMemInMib?

File: processor/memorylimiter/memorylimiter.go
Comment (bogdandrutu) [reply processor/memorylimiter/memorylimiter.go:312]: Not sure I understand the "This way, they wouldn't need to figure out where the config for this is located in their internal source repositories and which value should currently be used."? I understand the reasoning to expose this as metric or log, or even in zpages, but not sure I understand the part of looking into source code.

File: testbed/tests/trace_test.go
Comment (bogdandrutu) [inline testbed/tests/trace_test.go:169]: Why do you remove the jaeger and zipkin tests?

File: processor/memorylimiter/memorylimiter.go
Comment (jpkrohling) [reply processor/memorylimiter/memorylimiter.go:312]: Sorry, should have been clearer. People store configuration files in configuration management tools and/or source control mechanisms (git). The person looking at the memory consumption for the collector, perhaps via Prometheus or something even farther away, like a Grafana dashboard, might not readily have access to the configuration files for the collector. So, providing the configuration value as expvar/gauge would make their lives easier.

File: processor/memorylimiter/memorylimiter.go
Comment (jpkrohling) [reply processor/memorylimiter/memorylimiter.go:243]: Is `SwapInt64` a noop if the new value equals the old value?

File: processor/memorylimiter/README.md
Comment (tigrannajaryan) [reply processor/memorylimiter/README.md]: I added some recommendation for the relative sizes of soft vs hard. The absolute sizes do not belong to this doc, they should be part of the sizing recommendation (which we have elsewhere).

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go:243]: Reworked a bit. I don't want to use CAS or dawp, it makes the logic difficult to understand, especially since there is no bool version.

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go]: Simplified the code as suggested.

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go]: done.

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go]: renamed.

File: testbed/tests/trace_test.go
Comment (tigrannajaryan) [reply testbed/tests/trace_test.go:169]: They are not needed. We are not testing protocols here, we are testing memory limiter and testing using one protocol is sufficient, the behavior of memory limiter is not protocol dependent.
I cleaned this up more, PTAL.

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go:301]: We log the value of memSpikeLimit. A metric can be added in a future PR if we feel strong about it.

File: processor/memorylimiter/README.md
Comment (tigrannajaryan) [reply processor/memorylimiter/README.md:9]: I did another pass, hopefuly it is better now. Feel free to submit a PR with more fixes after this PR is merged.

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go]: I don't want to begin dropping sooner than we have to. GC has a good chance to free memory and avoid starting dropping. I would prefer to keep the logic this way.
In fact, I am thinking to invoke GC even when we pass the soft limit for the first time in the upward direction. Probably can be improved in a future PR.

Comment (bogdandrutu) [reply processor/memorylimiter/memorylimiter.go]: If you don't drop you have a big chance of OOM-ing at this moment, you are above the hard limit....

**Comment by tigrannajaryan** [comment]
Don't merge yet, I am looking into improving the limiting logic a bit more.

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go]: This should not happen if interval and spike settings are selected correctly. If they are selected incorrectly nothing will help.

**Comment by tigrannajaryan** [comment]
@jpkrohling @bogdandrutu I made some slight changes to the overall logic. We now try a GC when we hit a soft limit, but we limit the frequency of GC when we are between soft and hard limit. If GC helps and brings it down below soft limit we don't enable dropping. This helps to avoid data losses when we just touch the soft limit but there is plenty of garbage to be collected.

File: processor/memorylimiter/memorylimiter.go
Comment (jpkrohling) [inline processor/memorylimiter/memorylimiter.go:290]: Other than potentially calling `ml.setForcingDrop(mustForceDrop)`, are you doing anything in this function when the last GC happened less than 10s ago? If not, you might want to move this check to the top of the function, and return early.

**Review by jpkrohling — APPROVED** [review]
Just a minor comment, but LGTM already. Further concerns can be addressed based on real world usage, on follow-up PRs.

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go:290]: >are you doing anything in this function when the last GC happened less than 10s ago?

Yes, we may still do the GC if we are above hard limit. We also need to check the soft limit and change the drop flag, so I don't see how we can return early only by checking this condition.

**Review by jpkrohling — APPROVED** [review]

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go]: Updated to do GC earlier.

**Review by jrcamp — APPROVED** [review]

File: processor/memorylimiter/memorylimiter.go
Comment (jrcamp) [inline processor/memorylimiter/memorylimiter.go:261]: Might be useful to track as a custom metric 1) how long GC is taking to run 2) when is it running (unless I'm missing something that is already there).

Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go:261]: Yes. Can be added in a future PR.

File: processor/memorylimiter/memorylimiter.go
Comment (bogdandrutu) [inline processor/memorylimiter/memorylimiter.go:287]: I would move this next line (without else), I know it is an extra check but looks nicer :)

**Review by bogdandrutu — APPROVED** [review]

File: processor/memorylimiter/memorylimiter.go
Comment (tigrannajaryan) [reply processor/memorylimiter/memorylimiter.go]: Done.

**Pull request merged**
